website/server/libs/payments.js

```diff
@@ -321,4 +322,180 @@ api.buyGems = async function buyGems (data) {
+  let [
+    token,
+    user,
+    gift,
+    sub,
+    groupId,
+    email,
+    headers,
+  ] = options;
+  let coupon;
+  let response;
+  let subscriptionId;
+  // @TODO: We need to mock this, but curently we don't have correct Dependency Injection
+  let stripeApi = stripe;
+
+  if (stripeInc) stripeApi = stripeInc;
+
+  if (!token) throw new BadRequest('Missing req.body.id');
+
+  if (sub) {
+    if (sub.discount) {
+      if (!req.query.coupon) throw new BadRequest(res.t('couponCodeRequired'));
+      coupon = await Coupon.findOne({_id: cc.validate(req.query.coupon), event: sub.key});
+      if (!coupon) throw new BadRequest(res.t('invalidCoupon'));
+    }
+
+    let customerObject = {
+      email,
+      metadata: { uuid: user._id },
+      card: token,
+      plan: sub.key,
+    };
+
+    if (groupId) {
+      customerObject.quantity = sub.quantity;
+    }
+
+    response = await stripeApi.customers.create(customerObject);
+
+    if (groupId) subscriptionId = response.subscriptions.data[0].id;
+  } else {
+    let amount = 500; // $5
+
+    if (gift) {
+      if (gift.type === 'subscription') {
+        amount = `${shared.content.subscriptionBlocks[gift.subscription.key].price * 100}`;
+      } else {
+        amount = `${gift.gems.amount / 4 * 100}`;
+      }
+    }
+
+    response = await stripe.charges.create({
+      amount,
+      currency: 'usd',
+      card: token,
+    });
+  }
+
+  if (sub) {
+    await this.createSubscription({
+      user,
+      customerId: response.id,
+      paymentMethod: 'Stripe',
+      sub,
+      headers,
+      groupId,
+      subscriptionId,
+    });
+  } else {
+    let method = 'buyGems';
+    let data = {
+      user,
+      customerId: response.id,
+      paymentMethod: 'Stripe',
+      gift,
+    };
+
+    if (gift) {
+      let member = await User.findById(gift.uuid);
+      gift.member = member;
+      if (gift.type === 'subscription') method = 'createSubscription';
+      data.paymentMethod = 'Gift';
+    }
+
+    await this[method](data);
+  }
+};
+
+/**
+ * Allows for purchasing a user subscription or group subscription with Amazon
+ *
+ * @param  options
+ * @param  options.billingAgreementId  The Amazon billingAgreementId generated on the front end
+ * @param  options.user  The user object who is purchasing
+ * @param  options.sub  The subscription data to purchase
+ * @param  options.coupon  The coupon to discount the sub
+ * @param  options.groupId  The id of the group purchasing a subscription
+ * @param  options.headers  The request headers to store on analytics
+ * @return undefined
+ */
+api.subscribeWithAmazon = async function subscribeWithAmazon (options) {
+  let [
+    billingAgreementId,
+    sub,
+    coupon,
+    user,
+    groupId,
+    headers,
+  ] = options;
+
+  if (!sub) throw new BadRequest(res.t('missingSubscriptionCode'));
+  if (!billingAgreementId) throw new BadRequest('Missing req.body.billingAgreementId');
+
+  if (sub.discount) { // apply discount
```

Comment:
I'll abstract it when I write a test to cover it. I don't want to modify code without tests yet.